### PR TITLE
Replacement implementation for groupslices and associated functions

### PR DESCRIPTION
**Pull request by AndyGreenwell**
This PR replaces the original functionality proposed in issue #14142.

The main groupslices function returns the vector `uniquerow` (also referred to as `ic` in the prior discussion) as originally calculated by the `unique(A,dim)` implementation in multidimensional.jl.  A series of wrapper functions (`groupinds`, `firstinds`, `lastinds`) are included for returning the vector `ia` and vector of vectors `ib`, as discussed within #14142.

### CODE REVIEW

File: test/multidimensional.jl [hunk 1]
```diff
@@ -0,0 +1,30 @@
+# This file is a part of Julia. License is MIT: http://julialang.org/license
+
+using Base.Test
```
Comment (tkelman) [inline test/multidimensional.jl]: need to add this file to the list in `test/choosetests.jl` for it to run

File: base/multidimensional.jl [hunk 2]
```diff
@@ -931,3 +931,171 @@ If `dim` is specified, returns unique regions of the array `itr` along `dim`.
...
+
+Returns a vector of vectors of integers wherein the vector of group slice
+index integers as returned from `groupslices(A, dim)` is converted into a
+grouped vector of vectors.  Each vector entry in the returned vector of
+vectors contains all of the positional indices of slices in the original
+input array `A` that correspond to the unique slices along dimension `dim`
+that are present in the array `C` as returned from `unique(A, dim)`.
+"""
+
+function groupinds(ic::Vector{Int})
```
Comment (JoshChristie) [inline base/multidimensional.jl]: This doesn't seem to work in my use case. (I haven't switched to your branch or updated to 0.5. I just copied your code changes)
```julia
A =  [1.0  1.0  2.0  2.0  1.0  1.0  1.0  2.0  2.0  2.0
      1.0  1.0  2.0  2.0  2.0  1.0  2.0  2.0  2.0  1.0
      1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  2.0]
ic = groupslices(A, 2)
ib = groupinds(ic)
ERROR: BoundsError: attempt to access 4-element Array{Array{Int64,1},1}:
 [1,2]  
 Int64[]
 [3,4]  
 Int64[]
  at index [5]
in groupinds at 1046
```

My fix: 
```julia
function groupinds(ic::Vector{Int})
    d = Dict{Int, Int}()
    ia = unique(ic)
    n = length(ia)
    for i = 1:n
        d[ia[i]]= i
    end

    ib = Array(Vector{Int},n)
    for k = 1:n
        ib[k] = Int[]
    end

    for h = 1:length(ic)
        push!(ib[d[ic[i]]], h)
    end
    return ib
end
```

Comment (AndyGreenwell) [inline base/multidimensional.jl]: Thanks, using a Dict similar to `d` above is along the lines of what was in the original proposal [here](https://github.com/JuliaLang/julia/pull/14142/files#diff-89e97f67f8058d32eac8d917db5bbdbfR977) and [here](https://github.com/JuliaLang/julia/pull/14142/files#diff-89e97f67f8058d32eac8d917db5bbdbfR1045).  I'll incorporate your changes.

File: base/multidimensional.jl [hunk 3]
```diff
@@ -940,3 +940,178 @@ If `dim` is specified, returns unique regions of the array `itr` along `dim`.
...
+
+if dim == 1
+   all(A .== C[ic,:,:])
+elseif dim == 2
+   all(A .== C[:,ic,:])
+elseif dim == 3
+   all(A .== C[:,:,ic])
+end
+"""
+
```
Comment (tkelman) [inline base/multidimensional.jl]: remove the empty line between the docstring and the function it is documenting

File: base/multidimensional.jl [hunk 4]
```diff
@@ -940,3 +940,178 @@ If `dim` is specified, returns unique regions of the array `itr` along `dim`.
         @nref $N A d->d == dim ? sort!(uniquerows) : (1:size(A, d))
     end
 end
+
+"""
+    groupslices(A, dim)
```
Comment (tkelman) [inline base/multidimensional.jl]: see CONTRIBUTING.md, the signature line needs to be added to the rst docs, then run `make docs` to populate the docstring content

File: base/multidimensional.jl [hunk 5]
```diff
@@ -940,3 +940,178 @@ If `dim` is specified, returns unique regions of the array `itr` along `dim`.
...
+
+Returns a vector of integers where each integer element of the returned vector
+is a group number corresponding to the unique slices along dimension `dim` as
+returned from `unique(A, dim)`, where `A` can be a multidimensional array.
+
+Example usage:
+
+If `C = unique(A, dim)`, `ic = groupslices(A, dim)`, and
+rank(A) == rank(C) == 3, then:
```
Comment (tkelman) [inline base/multidimensional.jl]: should have backticks, and I think you mean `ndims` here not linear algebra rank

File: base/multidimensional.jl [hunk 6]
```diff
@@ -940,3 +940,178 @@ If `dim` is specified, returns unique regions of the array `itr` along `dim`.
...
+
+if dim == 1
```
Comment (tkelman) [inline base/multidimensional.jl]: code examples should be indented so the doc systen renders them as such

File: base/multidimensional.jl [hunk 7]
```diff
@@ -940,3 +940,178 @@ If `dim` is specified, returns unique regions of the array `itr` along `dim`.
...
+
+"""
+    groupinds(ic)
+
+Returns a vector of vectors of integers wherein the vector of group slice
+index integers as returned from `groupslices(A, dim)` is converted into a
+grouped vector of vectors.  Each vector entry in the returned vector of
+vectors contains all of the positional indices of slices in the original
+input array `A` that correspond to the unique slices along dimension `dim`
+that are present in the array `C` as returned from `unique(A, dim)`.
```
Comment (tkelman) [inline base/multidimensional.jl]: the wording here is tough to follow - should probably be rephrased

File: base/multidimensional.jl [hunk 8]
```diff
@@ -940,3 +940,178 @@ If `dim` is specified, returns unique regions of the array `itr` along `dim`.
...
+If `C = unique(A, dim)`, `ic = groupslices(A, dim)`, and
+rank(A) == rank(C) == 3, then:
+
+if dim == 1
+   all(A .== C[ic,:,:])
+elseif dim == 2
+   all(A .== C[:,ic,:])
+elseif dim == 3
+   all(A .== C[:,:,ic])
+end
```
Comment (JoshChristie) [inline base/multidimensional.jl]: This example usage fails for me. 

```julia
A =  [1.0  1.0  2.0  2.0  1.0  1.0  1.0  2.0  2.0  2.0
      1.0  1.0  2.0  2.0  2.0  1.0  2.0  2.0  2.0  1.0
      1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  2.0]
dim = 2
C = unique(A, dim)
ic = groupslices(A, dim)

if dim == 1
   all(A .== C[ic,:,:])
elseif dim == 2
   all(A .== C[:,ic,:])
elseif dim == 3
   all(A .== C[:,:,ic])
end

ERROR: BoundsError: attempt to access 3x4 Array{Float64,2}:
 1.0  2.0  1.0  2.0
 1.0  2.0  2.0  1.0
 1.0  1.0  1.0  2.0
  at index [Colon(),[1,1,3,3,5,1,5,3,3,10],Colon()]
 in throw_boundserror at abstractarray.jl:156
 in _internal_checkbounds at abstractarray.jl:176
 in checkbounds at abstractarray.jl:159
 in _getindex at multidimensional.jl:185
 in getindex at abstractarray.jl:488
```

File: base/multidimensional.jl (hunk 4 above)
Comment (AndyGreenwell) [inline base/multidimensional.jl]: In which *.rst file should the signature line for this function live?  At present, there is no multidimensional.rst file.

File: base/multidimensional.jl (hunk 5 above)
Comment (AndyGreenwell) [inline base/multidimensional.jl]: Will update.

File: base/multidimensional.jl (hunk 8 above)
Comment (AndyGreenwell) [inline base/multidimensional.jl]: Before I make another commit, I need to work get the documentation strings integrated properly as @tkelman points out, but the changes in this [gist](https://gist.github.com/AndyGreenwell/4c2f1481f5c974eb14f8c8266affe85b) should allow your example to work appropriately.  In one of the changes made in the course of discussing #14142 related to detecting hash collisions, where `ic` being directly assigned the values of `uniquerow` was improper.  While the values of `uniquerow` do contain the indices of unique slices of `A`, they are not the indices corresponding to the unique slices in `C` that allow for reconstruction of `A`.

The changes in this [gist](https://gist.github.com/AndyGreenwell/4c2f1481f5c974eb14f8c8266affe85b) assign the appropriate values to `ic` that allow one to reconstruct `A` by indexing `C` with `ic`, as had been the case in the original `uniqueslices` submission (but had not been handling hash collisions properly).

Thank you for testing.

**Comment by AndyGreenwell** [comment]
The functionality from this PR is now included in the following repository:

https://github.com/AndyGreenwell/GroupSlices.jl

That repo has been tagged, registered, and submitted as a [PR](https://github.com/JuliaLang/METADATA.jl/pull/5726) to METADATA.jl.  Once that PR is merged in METADATA, then I will close this PR.

**Pull request closed without merging**